pkg/apis/autoscaling/validation/validation.go

```diff
@@ -36,6 +37,33 @@ const (
 	MaxStabilizationWindowSeconds int32 = 3600
 )
 
+// HPAValidationOptions contains the different settings for HPA validation
+type HPAValidationOptions struct {
+	AllowMismatchedTargetTypeAndValue bool
+}
+
+// ValidationOptionsForHPA generates HPAValidationOptions for HorizontalPodAutoscaler
+func ValidationOptionsForHPA(newAutoscaler, oldAutoscaler *autoscaling.HorizontalPodAutoscaler) HPAValidationOptions {
+	opts := HPAValidationOptions{
+		AllowMismatchedTargetTypeAndValue: false,
+	}
+	// Don't allow mismatched target type/value fields for new HPAs
+	if oldAutoscaler == nil {
+		return opts
+	}
+	// Check for existing bad mismatched target type/value fields for ObjectMetricSource
+	// If UtilizationMetricType was set previously then either one of the values could have been used to pass validation before
+	for _, ms := range oldAutoscaler.Spec.Metrics {
+		if ms.Object != nil && ((ms.Object.Target.Type == autoscaling.ValueMetricType && ms.Object.Target.Value == nil) ||
+			(ms.Object.Target.Type == autoscaling.AverageValueMetricType && ms.Object.Target.AverageValue == nil) ||
+			(ms.Object.Target.Type == autoscaling.UtilizationMetricType)) {
+			opts.AllowMismatchedTargetTypeAndValue = true
+			return opts
+		}
+	}
```

Comment:
This bit would be easier to read in a standalone method that could return early:

```go
func hasMismatchedObjectMetricType(hpa *autoscaling.HorizontalPodAutoscaler) bool {
  if hpa == nil {
    return false
  }
  for _, ms := range oldAutoscaler.Spec.Metrics {
    switch {
    case ms.Object == nil:
      continue
    case ms.Object.Target.Type == autoscaling.ValueMetricType && ms.Object.Target.Value == nil:
      return true
    case ms.Object.Target.Type == autoscaling.AverageValueMetricType && ms.Object.Target.AverageValue == nil:
      return true
    case ms.Object.Target.Type == autoscaling.UtilizationMetricType:
      // If object metrics accept UtilizationMetricType in the future, expand this case to also check ms.Object.Target.AverageUtilizationValue == nil
      return true
    }
  }
  return false
}
```